File: Solution Heap.cpp

#include <array>
#include <queue>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {

    struct ScoreAndIndex {
        int score;
        size_t index;
        ScoreAndIndex(int score, size_t index): score{score}, index{index}{};
    };
    inline static const array<string, 3> topThreeRanks{ "Gold Medal", "Silver Medal", "Bronze Medal"};

public:
    vector<string> findRelativeRanks(const vector<int>& scores) const {
        static const auto comparator = [](const auto& first, const auto& second) {return first.score < second.score;};
        priority_queue<ScoreAndIndex, vector<ScoreAndIndex>, decltype(comparator) > maxHeap(comparator);
        for (size_t i = 0; i < scores.size(); ++i) {
            maxHeap.push(ScoreAndIndex(scores[i], i));
        }

        vector<string> allRanks(scores.size());
        int rank = 1;

        while (!maxHeap.empty() && rank < 4) {
            ScoreAndIndex current = maxHeap.top();
            maxHeap.pop();
            allRanks[current.index] = topThreeRanks[rank - 1];
            ++rank;
        }

        while (!maxHeap.empty()) {
            ScoreAndIndex current = maxHeap.top();
            maxHeap.pop();
            allRanks[current.index] = to_string(rank);
            ++rank;
        }

        return allRanks;
    }
};

File: Solution Heap.java

import java.util.PriorityQueue;

public class Solution {

    private record ScoreAndIndex(int score, int index) {}
    private static final String[] topThreeRanks = {"Gold Medal", "Silver Medal", "Bronze Medal"};

    public String[] findRelativeRanks(int[] scores) {
        PriorityQueue<ScoreAndIndex> maxHeap = new PriorityQueue<>((x, y) -> y.score - x.score);
        for (int i = 0; i < scores.length; ++i) {
            maxHeap.add(new ScoreAndIndex(scores[i], i));
        }

        String[] allRanks = new String[scores.length];
        int rank = 1;

        while (!maxHeap.isEmpty() && rank < 4) {
            ScoreAndIndex current = maxHeap.poll();
            allRanks[current.index] = topThreeRanks[rank - 1];
            ++rank;
        }

        while (!maxHeap.isEmpty()) {
            ScoreAndIndex current = maxHeap.poll();
            allRanks[current.index] = Integer.toString(rank);
            ++rank;
        }

        return allRanks;
    }
}
